### PR TITLE
Add postgres ARRAY type discovery and creation ability

### PR DESCRIPTION
**Pull request by cpcloud**
closes https://github.com/ContinuumIO/blaze/issues/1210

This significantly uglifies dshape_to_alchemy (which was already getting gross).
It might be time for that to be refactored before the next release

### CODE REVIEW

**Comment by rabernat** [comment]
:+1:

File: odo/backends/tests/test_postgres.py [hunk 1]
```diff
@@ -91,8 +92,24 @@ def complex_sql(url):
...
+            yield t
+        finally:
+            drop(t)
+
+
+@pytest.yield_fixture
+def sql_with_array(url):
+    ds = 'var * {a: var * float64, b: string}'
+    try:
+        t = resource(url, dshape=ds)
```
Comment (llllllllll) [inline odo/backends/tests/test_postgres.py]: do you think we should explicitly create the sql table here. If there is a bug where this doesn't actually create an array but is also not discovered as an array then the test will pass.

Comment (cpcloud) [inline odo/backends/tests/test_postgres.py]: yes that's a good idea, and i'll add a separate test for creating the table with odo

File: odo/backends/sql.py [hunk 2]
```diff
@@ -307,19 +317,34 @@ def dshape_to_alchemy(dshape):
...
+                              dshape_to_alchemy(typ, dialect, level=level + 1),
+                              nullable=isinstance(typ, Option))
+                    for name, typ in dshape.parameters[0]]
+        raise TypeError(
+            'nested record dshape not supported by sqlalchemy')
     if isinstance(dshape, datashape.DataShape):
         if isdimension(dshape[0]):
-            return dshape_to_alchemy(dshape[1])
+            if isscalar(dshape[1]):
+                type_ = array_types.get(dialect.name, None)
```
Comment (llllllllll) [inline odo/backends/sql.py]: `None` is the default arg here

Comment (cpcloud) [inline odo/backends/sql.py]: Yeah I'll move it up so that the `dialect` appears as `dialect.name`

File: odo/backends/sql.py [hunk 3]
```diff
@@ -307,19 +317,34 @@ def dshape_to_alchemy(dshape):
...
+                if type_ is None:
+                    raise TypeError('dialect %s has no array type for dshape '
+                                    '%s' % (dialect.name, dshape))
+                return type_(dshape_to_alchemy(dshape[1],
+                                               dialect, level=level + 1))
+            else:
+                if not level:
+                    return dshape_to_alchemy(dshape[1], dialect,
+                                             level=level + 1)
+                raise TypeError('nested record dshape not supported by sqlalchemy')
```
Comment (llllllllll) [inline odo/backends/sql.py]: Is this going to make it more difficult to implement your foreign key feature?

Comment (cpcloud) [inline odo/backends/sql.py]: Yes, thanks for pointing that out. I'll merge this into that and see what happens.

**Comment by llllllllll** [comment]
If someone wants to own this please reopen

**Pull request closed without merging**